src/validation.rs

```diff
@@ -270,24 +272,44 @@ impl Validator {
                 ..
             } => {
                 self.visit_statement(selector, context);
+
+                let mut cases = HashSet::new();
                 case_blocks.iter().for_each(|b| {
                     let condition = b.condition.as_ref();
-                    if let AstStatement::Reference { name, .. } = condition {
-                        if let Some(StatementAnnotation::Variable {
-                            constant: false, ..
-                        }) = context.ast_annotation.get(condition)
-                        {
+
+                    // validate for duplicate conditions
+                    // first try to evaluate the conditions value
+                    const_evaluator::evaluate(condition, context.qualifier, context.index)
+                        .map_err(|err| {
+                            // value evaluation and validation not possible with non constants
                             self.stmt_validator.diagnostics.push(
                                 Diagnostic::non_constant_case_condition(
-                                    name,
+                                    &err,
                                     condition.get_location(),
                                 ),
-                            );
-                        }
-                    }
+                            )
+                        })
+                        .map(|v| {
+                            // check for duplicates if we got a value
+                            if let Some(AstStatement::LiteralInteger { value, .. }) = v {
+                                if cases.contains(&value) {
```

Comment:
insert in a set returns false if the value already exists, so you can do both check and insert in one go `if !case.insert(&value)`